todo-list-backend/tests/integration/graphql-api.spec.ts

```diff
@@ -0,0 +1,526 @@
+import {buildApolloServer} from "../../src";
+import {InMemoryStorage, IStorage} from "../../src/data";
+import {User, ToDo} from "../../src/types";
+import {ApolloServerTestClient, createTestClient} from "apollo-server-testing";
+import {gql} from "apollo-server";
+import {sign} from "jsonwebtoken";
+
+const READ_TODOS = gql`query { readTodos { id name done assignee { name }}} `;
+
+const LOGIN = gql`
+    mutation Login($name: String!, $password: String!) {
+        login(name: $name, password: $password)
+    }
+`;
+
+const CREATE_TODO = gql`
+    mutation CreateTodo($name: String!) {
+        createTodo(name: $name) {
+            id name done assignee { name }
+        }
+    }
+`;
+
+const UPDATE_TODO_NAME = gql`
+    mutation UpdateTodoName($id: Int!, $name: String!) {
+        updateTodoName(id: $id, name: $name) {
+            id name done assignee { name }
+        }
+    }
+`;
+
+const UPDATE_TODO_DONE = gql`
+    mutation UpdateTodoDone($id: Int!, $done: Boolean!) {
+        updateTodoDone(id: $id, done: $done)  {
+            id name done assignee { name }
+        }
+    }
+`;
+
+const DELETE_TODO = gql`
+    mutation DeleteTodo($id: Int!) {
+        deleteTodo(id: $id)  {
+            id name done assignee { name }
+        }
+    }
+`;
+
+const CREATE_USER = gql`
+    mutation CreateUser($name: String!, $password: String!) {
+        createUser(name: $name, password: $password) {
+            name
+        }
+    }
+`;
+
+function buildTwoUserInMemoryStorage(addTodos: boolean = false): IStorage {
+    let initialUsers = [
+        new User("ralph", "ralph"),
+        new User("nora", "nora")
+    ];
+
+    if (!addTodos) {
+        return new InMemoryStorage(initialUsers);
+    }
+
+    let initialTodos = [
+        new ToDo("first todo", initialUsers[1]),
+        new ToDo("second todo", initialUsers[0]),
+    ];
+
+    return new InMemoryStorage(initialUsers, initialTodos);
+
+}
+
+describe("query", () => {
+    const authSecret: string = "This is a perfect secret for testing!";
+    let user: User;
+    let storage: IStorage;
+    let query: any;
+    let mutate: any;
+
+
+    describe("given a legit authorization token", () => {
+        beforeEach(async () => {
+            storage = buildTwoUserInMemoryStorage(true);
+
+            // workaround for the shortcomings of the apollo-server-testing
+            // package as it doesn't generate a real request and therefor a
+            // "real" context will never be generated
+            user = storage.readUser("ralph");
+            const token = sign({username: user.name}, authSecret, {expiresIn: "1 day"});
+            const context = {token: token, user: user};
```

Comment:
That's not even a workaround, it's perfectly fine. In our backend tests at Human Connection we mock the `context` method like that. It has the currently authenticated user so we can vary the user who issues the request in each test case.